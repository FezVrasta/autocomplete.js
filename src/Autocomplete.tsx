/** @jsx h */

import { h } from 'preact';
import { useState, useEffect, useRef } from 'preact/hooks';
import { createPortal } from 'preact/compat';
import Downshift from 'downshift/preact';

import { flatten, noop } from './utils';
import { Dropdown } from './Dropdown';
import { SearchBox } from './SearchBox';
import {
  Environment,
  AutocompleteProps,
  AutocompleteState,
  AutocompleteSource,
  AutocompleteItem,
} from './types';

export const defaultEnvironment =
  typeof window === 'undefined' ? ({} as Environment) : window;

let autocompleteIdCounter = 0;

/**
 * Generates a unique ID for an instance of an Autocomplete DownShift instance.
 */
function generateId(): string {
  return String(autocompleteIdCounter++);
}

export function Autocomplete(props: AutocompleteProps) {
  const {
    container,
    placeholder = '',
    minLength = 1,
    showCompletion = false,
    autofocus = false,
    initialState = {},
    defaultHighlightedIndex = 0,
    stalledDelay = 300,
    keyboardShortcuts = [],
    getSources,
    templates = {},
    environment = defaultEnvironment,
    dropdownContainer = environment.document.body,
    dropdownPosition = 'left',
    onFocus = noop,
    onClick = noop,
    onKeyDown = noop,
    onEmpty = noop,
    onError = ({ state }) => {
      throw state.error;
    },
  } = props;

  const [query, setQuery] = useState<AutocompleteState['query']>(
    initialState.query || ''
  );
  const [results, setResults] = useState<AutocompleteState['results']>(
    initialState.results || []
  );
  const [isOpen, setIsOpen] = useState<AutocompleteState['isOpen']>(
    initialState.isOpen || false
  );
  const [isLoading, setIsLoading] = useState<AutocompleteState['isLoading']>(
    initialState.isLoading || false
  );
  const [isStalled, setIsStalled] = useState<AutocompleteState['isStalled']>(
    initialState.isStalled || false
  );
  const [error, setError] = useState<AutocompleteState['error'] | null>(
    initialState.error || null
  );
  const [metadata, setMetadata] = useState<AutocompleteState['metadata']>(
    initialState.metadata || {}
  );
  const [sources, setSources] = useState<AutocompleteSource[]>(
    getSources({ query })
  );
  const [dropdownRect, setDropdownRect] = useState<
    Pick<ClientRect, 'top' | 'left'> | undefined
  >(undefined);

  let setIsStalledId: number | null = null;
  const inputRef = useRef<HTMLInputElement | null>(null);

  useEffect(() => {
    // Perform the query if coming from the initial state.
    if (initialState.query) {
      performQuery(initialState.query, isOpen);
    }
  }, []);

  useEffect(() => {
    function onGlobalKeyDown(event: KeyboardEvent): void {
      if (!inputRef.current) {
        return;
      }

      const element = event.target as HTMLElement;
      const tagName = element.tagName;

      // Do not trigger the focus if we're already editing text.
      if (
        element.isContentEditable ||
        tagName === 'INPUT' ||
        tagName === 'SELECT' ||
        tagName === 'TEXTAREA'
      ) {
        return;
      }

      // Do not trigger the focus if the shortcut is not correct.
      if (keyboardShortcuts.indexOf(event.key) === -1) {
        return;
      }

      inputRef.current.focus();

      event.stopPropagation();
      event.preventDefault();
    }

    if (keyboardShortcuts.length > 0) {
      environment.addEventListener('keydown', onGlobalKeyDown);
    }

    return () => {
      if (keyboardShortcuts.length > 0) {
        environment.removeEventListener('keydown', onGlobalKeyDown);
      }
    };
  }, [environment, keyboardShortcuts]);

  useEffect(() => {
    environment.addEventListener('resize', onResize);

    return () => {
      environment.removeEventListener('resize', onResize);
    };
  }, [environment, onResize]);

  useEffect(() => {
    // We need to track the container position because the dropdown position is
    // computed based on the container position.
    onResize();
  }, [container, dropdownContainer, onResize]);

  function getState(): AutocompleteState {
    return {
      query,
      results,
      isOpen,
      isLoading,
      isStalled,
      error,
      metadata,
    };
  }

  function setState(nextState: Partial<AutocompleteState>): void {
    if (nextState.query !== undefined) {
      setQuery(nextState.query);
      performQuery(nextState.query);
    }
    if (nextState.results) {
      setResults(nextState.results);
    }
    if (nextState.isOpen !== undefined) {
      setIsOpen(nextState.isOpen);
    }
    if (nextState.isLoading !== undefined) {
      setIsLoading(nextState.isLoading);
    }
    if (nextState.isStalled !== undefined) {
      setIsStalled(nextState.isStalled);
    }
    if (nextState.error !== undefined) {
      setError(nextState.error);
    }
    if (nextState.metadata !== undefined) {
      setMetadata({ ...metadata, ...nextState.metadata });
    }
  }

  function onResize(): void {
    const nextContainerRect = container.getBoundingClientRect();
    const nextDropdownRect = dropdownContainer.getBoundingClientRect();

    const newDropdownRect = {
      top: nextContainerRect.top + nextContainerRect.height,
      left:
        dropdownPosition === 'right'
          ? nextContainerRect.width +
            nextContainerRect.left -
            nextDropdownRect.width
          : nextContainerRect.left,
    };

    setDropdownRect(newDropdownRect);
  }

  function performQuery(query: string, nextIsOpen: boolean = true) {
    if (setIsStalledId) {
      clearTimeout(setIsStalledId);
      setIsStalled(false);
    }

    setError(null);
    setQuery(query);

    if (query.length < minLength) {
      setIsLoading(false);
      setIsOpen(false);
      setSources(getSources({ query }));
      setResults([]);

      return Promise.resolve();
    }

    setIsLoading(true);

    setIsStalledId = environment.setTimeout(() => {
      setIsStalled(true);
    }, stalledDelay);

    return Promise.all(
      sources.map(source => {
        if (source.onInput) {
          source.onInput({ state: getState(), setState });
        }

        return source.getSuggestions({
          query,
          state: getState(),
          setState,
        });
      })
    )
      .then(results => {
        if (setIsStalledId) {
          clearTimeout(setIsStalledId);
          setIsStalled(false);
        }

        setIsLoading(false);
        setIsOpen(nextIsOpen);
        setResults(results);
        setSources(getSources({ query }));

        const hasResults = results.some(result => result.length > 0);

        if (!hasResults) {
          onEmpty({ state: getState(), setState });
        }
      })
      .catch(error => {
        if (setIsStalledId) {
          clearTimeout(setIsStalledId);
          setIsStalled(false);
        }

        setIsLoading(false);
        setError(error);

        onError({
          state: getState(),
          setState,
        });
      });
  }

  function getSourceFromHighlightedIndex(
    highlightedIndex: number
  ): AutocompleteSource {
    const resultsSizes = results.reduce<number[]>((acc, result: number[]) => {
      acc.push(result.length + acc.reduce((a, b) => a + b, 0));
      return acc;
    }, []);
    const sourceNumber = resultsSizes.reduce((acc, current) => {
      if (current <= highlightedIndex) {
        return acc + 1;
      }

      return acc;
    }, 0);

    return sources[sourceNumber];
  }

  function getCompletion(highlightedIndex: number) {
    if (!showCompletion) {
      return '';
    }

    const suggestion = flatten(results)[Math.max(0, highlightedIndex)];
    const source = getSourceFromHighlightedIndex(Math.max(0, highlightedIndex));

    if (!suggestion || !source) {
      return '';
    }

    const currentSuggestion = source.getSuggestionValue({
      suggestion,
      state: getState(),
    });

    if (
      query &&
      currentSuggestion
        .toLocaleLowerCase()
        .indexOf(query.toLocaleLowerCase()) === 0
    ) {
      return query + currentSuggestion.slice(query.length);
    }

    return '';
  }

  const canOpen = query.length >= minLength;
  const hasResults = results.some(result => result.length > 0);
  const shouldOpen =
    isOpen &&
    // We don't want to open the dropdown when the results
    // are loading coming from an empty input.
    // !isStalled &&
    // However, we do want to leave the dropdown open when it's
    // already open because there are results displayed. Otherwise,
    // it would result in a flashy behavior.
    canOpen &&
    hasResults;

  return (
    <Downshift
      id={`autocomplete-${generateId()}`}
      environment={environment}
      defaultHighlightedIndex={defaultHighlightedIndex}
      onSelect={(item: AutocompleteItem) => {
        if (!item) {
          return;
        }

        const { suggestion, source } = item;

        performQuery(
          source.getSuggestionValue({ suggestion, state: getState() }),
          false
        ).then(() => {
          if (source.onSelect) {
            const currentState = getState();

            source.onSelect({
              suggestion,
              suggestionValue: source.getSuggestionValue({
                suggestion,
                state: currentState,
              }),
              source,
              state: currentState,
              setState,
            });
          } else {
            setIsOpen(false);
          }
        });
      }}
      onOuterClick={() => {
        setIsOpen(false);
      }}
      scrollIntoView={(itemNode: HTMLElement) => {
        if (itemNode) {
          itemNode.scrollIntoView(false);
        }
      }}
    >
      {({
        highlightedIndex,
        setHighlightedIndex,
        getInputProps,
        getItemProps,
        getMenuProps,
      }) => {
        return (
          <div
            className={[
              'algolia-autocomplete',
              isStalled && 'algolia-autocomplete--stalled',
              error && 'algolia-autocomplete--errored',
            ]
              .filter(Boolean)
              .join(' ')}
          >
            <SearchBox
              placeholder={placeholder}
              autofocus={autofocus}
              completion={getCompletion(highlightedIndex)}
              internalState={getState()}
              internalSetState={setState}
              onInputRef={inputRef}
              getInputProps={getInputProps}
              onFocus={() => {
                if (canOpen) {
                  setIsOpen(true);
                }

                if (minLength === 0 && !query) {
                  performQuery('');
                }

                onFocus({
                  state: getState(),
                  setState,
                });
              }}
              onKeyDown={(event: KeyboardEvent) => {
                const suggestion: any = flatten(results)[
                  Math.max(0, highlightedIndex)
                ];
                const source = getSourceFromHighlightedIndex(
                  Math.max(0, highlightedIndex)
                );

                const currentState = getState();

                if (suggestion && source) {
                  onKeyDown(event, {
                    suggestion,
                    suggestionValue: source.getSuggestionValue({
                      suggestion,
                      state: currentState,
                    }),
                    source,
                    state: currentState,
                    setState,
                  });
                } else {
                  onKeyDown(event, { state: currentState, setState });
                }

                if (event.key === 'Escape') {
                  setIsOpen(false);

                  if (!shouldOpen) {
                    setQuery('');
                  }
                } else if (
                  event.key === 'Tab' ||
                  // When the user hits the right arrow and is at the end of
                  // the typed query only, we want to validate the completion.
                  (event.key === 'ArrowRight' &&
                    (event.target as HTMLInputElement).selectionStart ===
                      query.length)
                ) {
                  if (shouldOpen && suggestion && source) {
                    event.preventDefault();

                    const nextQuery = source.getSuggestionValue({
                      suggestion,
                      state: getState(),
                    });

                    if (query !== nextQuery) {
                      performQuery(nextQuery);

                      setHighlightedIndex(defaultHighlightedIndex);
                    }
                  }
                }
              }}
              onInput={(event: KeyboardEvent) => {
                performQuery((event.target as HTMLInputElement).value);
              }}
              onReset={event => {
                event.preventDefault();
                setQuery('');

                if (minLength === 0) {
                  performQuery('');
                }

                if (inputRef.current) {
                  inputRef.current.focus();
                }
              }}
              onSubmit={(event: Event) => {
                event.preventDefault();
                setIsOpen(false);

                if (inputRef.current) {
                  inputRef.current.blur();
                }
              }}
            />

            {createPortal(
              <Dropdown
                position={dropdownRect}
                hidden={!shouldOpen}
                internalState={getState()}
                internalSetState={setState}
                sources={sources}
                templates={templates}
                onClick={onClick}
                getMenuProps={getMenuProps}
                getItemProps={getItemProps}
              />,
              dropdownContainer
            )}
          </div>
        );
      }}
    </Downshift>
  );
}
