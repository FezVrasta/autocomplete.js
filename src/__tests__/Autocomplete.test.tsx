/** @jsx h */

import { h } from 'preact';
import { render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';

import { Autocomplete } from '../Autocomplete';

function createEnvironment() {
  return {
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    setTimeout: jest.fn(),
    document: window.document,
  };
}

const defaultSource = {
  getSuggestionValue: ({ suggestion }) => suggestion.value,
  templates: {
    suggestion({ suggestion }) {
      return suggestion.value;
    },
  },
};

const fruits = [
  { value: 'Apple' },
  { value: 'Banana' },
  { value: 'Cherry' },
  { value: 'Orange' },
  { value: 'Pear' },
  { value: 'Pineapple' },
  { value: 'Raspberry' },
  { value: 'Strawberry' },
];

function getDefaultProps() {
  const environment = createEnvironment();

  return {
    container: environment.document.body,
    environment,
    getSources: () => [
      {
        ...defaultSource,
        getSuggestions() {
          return fruits;
        },
      },
    ],
  };
}

describe('Autocomplete', () => {
  test('should generated the correct DOM', () => {
    const props = {
      ...getDefaultProps(),
    };

    const { container } = render(<Autocomplete {...props} />);

    expect(container).toMatchSnapshot();
  });

  describe('minLength', () => {
    test.skip('should open the menu when the number of characeters is exceeded', () => {
      const props = {
        ...getDefaultProps(),
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );
      const dropdown = props.environment.document.body.querySelector<
        HTMLElement
      >('.algolia-autocomplete-dropdown');

      expect(dropdown).toHaveAttribute('hidden');

      userEvent.type(input, 'que');

      expect(dropdown).toHaveAttribute('hidden', 'false');
    });

    test('should not open the menu when the number of characeters is not exceeded', () => {
      const props = {
        ...getDefaultProps(),
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );
      const dropdown = props.environment.document.body.querySelector<
        HTMLElement
      >('.algolia-autocomplete-dropdown');

      input.focus();

      expect(dropdown).toHaveAttribute('hidden');
    });
  });

  describe('stalledDelay', () => {
    test.skip('should display information about the experience being stalled', () => {
      const props = {
        ...getDefaultProps(),
        stalledDelay: 0,
        getSources: () => [
          {
            ...defaultSource,
            getSuggestions() {
              return new Promise<object[]>(resolve => {
                setTimeout(() => {
                  resolve(fruits);
                }, 100);
              });
            },
          },
        ],
      };

      const { container } = render(<Autocomplete {...props} />);
      const stalledRoot = container.querySelector<HTMLElement>(
        '.algolia-autocomplete--stalled'
      );
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'query');

      expect(stalledRoot).toBeInTheDocument();
    });

    test('should not display information about the experience being stalled above delay', () => {
      const props = {
        ...getDefaultProps(),
        stalledDelay: 0,
      };

      const { container } = render(<Autocomplete {...props} />);
      const stalledRoot = container.querySelector<HTMLElement>(
        '.algolia-autocomplete--stalled'
      );
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'query');

      expect(stalledRoot).not.toBeInTheDocument();
    });
  });

  describe('dropdownContainer', () => {
    test.todo('should be the document body by default');
    test.todo('can be any other HTML element');
    test.todo('can be any other string referencing to an HTML element');
  });

  describe('dropdownPosition', () => {
    test.todo('defaults to `"left"`');
    test.todo('can be set to `"right"`');
  });

  describe('placeholder', () => {
    test.todo('is empty by default');
    test.todo('is forwarded to the input');
  });

  describe('defaultHighlightedIndex', () => {
    test.todo('is the first item by default');
    test.todo('can be other items');
    test.todo('can be negative to not select any items');
  });

  describe('autofocus', () => {
    test.todo('should not focus the input by default');
    test.todo('should focus the input when `true`');
  });

  describe('showCompletion', () => {
    test.todo('should be disable by default');
    test.todo('should start showing when typing');
  });

  describe('initialState', () => {
    test('allows to set the initial query', () => {
      const props = {
        ...getDefaultProps(),
        initialState: {
          query: 'Initial query',
        },
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      expect(input).toHaveValue('Initial query');
    });

    test.skip('allows to set the initial isOpen', () => {
      const props = {
        ...getDefaultProps(),
        initialState: {
          isOpen: true,
        },
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      const dropdown = props.environment.document.body.querySelector<
        HTMLElement
      >('.algolia-autocomplete-dropdown');

      expect(dropdown).toHaveAttribute('hidden', 'false');
    });

    test('allows to set the initial isStalled', () => {
      const props = {
        ...getDefaultProps(),
        initialState: {
          isStalled: true,
        },
      };

      const { container } = render(<Autocomplete {...props} />);
      const stalledContainer = container.querySelector(
        '.algolia-autocomplete--stalled'
      );

      expect(stalledContainer).toBeInTheDocument();
    });

    test.todo('allows to set the initial isLoading');

    test('allows to set the initial error', () => {
      const props = {
        ...getDefaultProps(),
        initialState: {
          error: new Error('Test error'),
        },
      };

      const { container } = render(<Autocomplete {...props} />);
      const erroredContainer = container.querySelector(
        '.algolia-autocomplete--errored'
      );

      expect(erroredContainer).toBeInTheDocument();
    });

    test.skip('allows to set the initial results', () => {
      const props = {
        ...getDefaultProps(),
        getSources() {
          return [
            {
              getSuggestionValue: () => '',
              getSuggestions: () => fruits,
              templates: {
                suggestion: () => '',
              },
            },
          ];
        },
        initialState: {
          results: [fruits],
        },
      };

      const { container } = render(<Autocomplete {...props} />);
    });

    test.todo('allows to set the initial metadata');
  });

  describe('keyboardShortcuts', () => {
    test('attaches keydown event to environment', () => {
      const props = {
        ...getDefaultProps(),
        keyboardShortcuts: ['/'],
      };

      const { unmount } = render(<Autocomplete {...props} />);

      expect(props.environment.addEventListener).toHaveBeenCalledWith(
        'keydown',
        expect.any(Function)
      );

      unmount();

      expect(props.environment.removeEventListener).toHaveBeenCalledWith(
        'keydown',
        expect.any(Function)
      );
    });

    test('does not attach keydown event to environment without keyboardShortcuts', () => {
      const props = {
        ...getDefaultProps(),
      };

      const { unmount } = render(<Autocomplete {...props} />);

      expect(props.environment.addEventListener).not.toHaveBeenCalledWith(
        'keydown',
        expect.any(Function)
      );

      unmount();

      expect(props.environment.addEventListener).not.toHaveBeenCalledWith(
        'keydown',
        expect.any(Function)
      );
    });
  });

  describe('onFocus', () => {
    test('is called when the input is focused', () => {
      const props = {
        ...getDefaultProps(),
        onFocus: jest.fn(),
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      expect(props.onFocus).toHaveBeenCalledTimes(0);

      input.focus();

      expect(props.onFocus).toHaveBeenCalledTimes(1);
      expect(props.onFocus).toHaveBeenCalledWith({
        state: expect.any(Object),
        setState: expect.any(Function),
      });

      input.blur();
      input.focus();

      expect(props.onFocus).toHaveBeenCalledTimes(2);
    });
  });

  describe('onClick', () => {
    test.skip('is called when an item is cliked', () => {
      const props = {
        ...getDefaultProps(),
        onClick: jest.fn(),
        minLength: 0,
        initialState: {
          isOpen: true,
          query: 'Apple',
          results: [fruits],
        },
      };

      const { container, debug } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'Apple');

      const [firstItem] = container.querySelectorAll<HTMLElement>(
        '.algolia-autocomplete-item'
      );

      firstItem.click();

      expect(props.onClick).toHaveBeenCalledTimes(1);
    });
  });

  describe('onKeyDown', () => {
    test('is forwarded to the input without selected item', () => {
      const props = {
        ...getDefaultProps(),
        onKeyDown: jest.fn(),
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'Query');

      expect(props.onKeyDown).toHaveBeenCalledTimes(5);
      expect(props.onKeyDown).toHaveBeenLastCalledWith(
        expect.any(KeyboardEvent),
        {
          state: expect.any(Object),
          setState: expect.any(Function),
        }
      );
    });

    test('is forwarded to the input with a selected item', () => {
      const props = {
        ...getDefaultProps(),
        onKeyDown: jest.fn(),
        getSources: () => [
          {
            ...defaultSource,
            getSuggestions() {
              return [];
            },
          },
        ],
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'Query');

      expect(props.onKeyDown).toHaveBeenCalledTimes(5);
      expect(props.onKeyDown).toHaveBeenLastCalledWith(
        expect.any(KeyboardEvent),
        {
          state: expect.any(Object),
          setState: expect.any(Function),
        }
      );
    });
  });

  describe.skip('onEmpty', () => {
    test('is called when there are no results', () => {
      const props = {
        ...getDefaultProps(),
        onEmpty: jest.fn(),
        getSources: () => [
          {
            ...defaultSource,
            getSuggestions({ query }) {
              return [{ value: 'Apple' }].filter(
                suggestion => suggestion.value === query
              );
            },
          },
        ],
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'Appl');

      expect(props.onEmpty).toHaveBeenCalledTimes(4);

      userEvent.type(input, 'Apple');

      expect(props.onEmpty).toHaveBeenCalledTimes(4);
    });
  });

  describe.skip('onError', () => {
    test('is called when fetching the suggestions fails', () => {
      const testError = new Error('Test error');
      const props = {
        ...getDefaultProps(),
        onError: jest.fn(({ state }) => {
          throw state.error;
        }),
        getSources: () => [
          {
            ...defaultSource,
            getSuggestions() {
              throw testError;
            },
          },
        ],
      };

      const { container } = render(<Autocomplete {...props} />);
      const input = container.querySelector<HTMLInputElement>(
        '.algolia-autocomplete-input'
      );

      userEvent.type(input, 'q');

      expect(props.onError).toHaveBeenCalledTimes(1);
      expect(props.onError).toHaveBeenCalledWith({
        state: expect.objectContaining({
          error: testError,
        }),
        setState: expect.any(Function),
      });
    });
  });
});
