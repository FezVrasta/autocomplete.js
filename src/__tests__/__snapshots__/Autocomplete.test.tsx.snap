// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should generated the correct DOM 1`] = `
<div>
  <div
    aria-haspopup="listbox"
    aria-labelledby="autocomplete-0-label"
    class="algolia-autocomplete"
    role="combobox"
  >
    <form
      action=""
      class="algolia-autocomplete-form"
      novalidate=""
      role="search"
    >
      <label
        class="algolia-autocomplete-magnifierLabel"
        for="autocomplete-0-input"
      >
        <svg
          viewBox="0 0 18 18"
        >
          <path
            d="M13.14 13.14L17 17l-3.86-3.86A7.11 7.11 0 1 1 3.08 3.08a7.11 7.11 0 0 1 10.06 10.06z"
            fill="none"
            fillRule="evenodd"
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="1.78"
          />
        </svg>
      </label>
      <div
        class="algolia-autocomplete-loadingIndicator"
      >
        <svg
          stroke="currentColor"
          strokeOpacity=".5"
          viewBox="0 0 38 38"
        >
          <g
            fill="none"
            fillRule="evenodd"
          >
            <g
              strokeWidth="2"
              transform="translate(1 1)"
            >
              <circle
                cx="18"
                cy="18"
                r="18"
                strokeOpacity=".3"
              />
              <path
                d="M36 18c0-9.94-8.06-18-18-18"
              >
                <animatetransform
                  attributeName="transform"
                  dur="1s"
                  from="0 18 18"
                  repeatCount="indefinite"
                  to="360 18 18"
                  type="rotate"
                />
              </path>
            </g>
          </g>
        </svg>
      </div>
      <div
        class="algolia-autocomplete-searchbox"
      >
        <input
          aria-autocomplete="list"
          aria-labelledby="autocomplete-0-label"
          autocapitalize="off"
          autocomplete="off"
          autocorrect="off"
          class="algolia-autocomplete-input"
          id="autocomplete-0-input"
          maxlength="512"
          placeholder=""
          spellcheck="false"
          type="search"
        />
      </div>
      <button
        class="algolia-autocomplete-reset"
        hidden=""
        title="Clear the query"
        type="reset"
      >
        <svg
          viewBox="0 0 10 10"
        >
          <path
            d="M5 4.12L8.93.18a.62.62 0 1 1 .89.89L5.88 5l3.94 3.93a.62.62 0 1 1-.89.89L5 5.88 1.07 9.82a.62.62 0 1 1-.89-.89L4.12 5 .18 1.07a.62.62 0 1 1 .89-.89L5 4.12z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
      </button>
    </form>
  </div>
</div>
`;
